src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultHtmlFactsService.cs

```diff
@@ -11,23 +11,29 @@ internal class DefaultHtmlFactsService : HtmlFactsService
     {
         public override bool TryGetElementInfo(SyntaxNode element, out SyntaxToken containingTagNameToken, out SyntaxList<RazorSyntaxNode> attributeNodes)
         {
-            if (element is MarkupStartTagSyntax startTag)
+            switch (element)
             {
-                containingTagNameToken = startTag.Name;
-                attributeNodes = startTag.Attributes;
-                return true;
-            }
-
-            if (element is MarkupTagHelperStartTagSyntax startTagHelper)
-            {
-                containingTagNameToken = startTagHelper.Name;
-                attributeNodes = startTagHelper.Attributes;
-                return true;
+                case MarkupStartTagSyntax startTag:
+                    containingTagNameToken = startTag.Name;
+                    attributeNodes = startTag.Attributes;
+                    return true;
+                case MarkupEndTagSyntax endTag:
+                    containingTagNameToken = endTag.Name;
+                    attributeNodes = ((MarkupElementSyntax)endTag.Parent).StartTag.Attributes;
```

Comment:
For safety's sake it would be better to not use casts here, but expand the pattern on the `case` blocks. So for example `case MarkupEndTagSyntax { Parent: MarkupElementSyntax parent } endTag:` and then this line would just be `attributeNodes = parent.StartTag.Attributes;`.

It reads a bit nicer, and if the tree ever gets wacky it won't throw an InvalidCastException, it will just fall through to the default case.

Same goes for `MarkupTagHelperEndTagSyntax` below

Reply:
Done